lib/link_checker.rb

```diff
@@ -0,0 +1,73 @@
+class LinkChecker
+  class << self
+    def extract_urls(data)
+      new(data).extract_urls
+    end
+
+    def link_checker_api_bearer_token
+      ENV["LINK_CHECKER_API_BEARER_TOKEN"]
+    end
+
+    def for_files_at(path)
+      path = File.join(path, "**/*.erb") unless path.to_s.include?("*")
+      data = Dir.glob(path).map { |filename| File.read(filename) }
+      new(data)
+    end
+  end
+
+  attr_reader :data
+
+  def initialize(data)
+    @data = data
+  end
+
+  def urls
+    # Gather urls in a Set to allow simple combination and duplicate removal
+    @urls ||= extract_urls.to_set
+  end
+
+  def extract_urls
+    case data
+    when String
+      uri = URI.extract(data, %w[http https])
+
+      # markdown urls trailing bracket included in uri so removing here
+      uri.map { |u| unmatched_closing_bracket?(u) ? u.gsub(/\)\W?$/, "") : u }
+    when Array
+      urls = data.map do |datum|
+        self.class.extract_urls(datum)
+      end
+      urls.flatten!
+      urls
+    else
+      []
+    end
+  end
+
+  def unmatched_closing_bracket?(text)
+    text.count("(") < text.count(")")
+  end
+
+  def report
+    @report ||= generate_report
+  end
+
+  def generate_report
+    link_report = link_checker.create_batch(urls, checked_within: 5)
+    wait_time = 0.5
+
+    while link_report.status == :in_progress
+      sleep(wait_time)
+      link_report = link_checker.get_batch(link_report.id)
+      wait_time *= 2
+    end
+    link_report
+  end
+
+  def link_checker
```

Comment:
Nit: 
Found this confusing as to whether it was returning a LinkChecker object or client for the Link Checker API
```suggestion
  def link_checker_api
```